DEVELOPMENT.md

```diff
@@ -0,0 +1,115 @@
+- False sharing/data locking protection issues, e.g.
+```go
+    type BeaconHandler struct {
+        // NOTE: should only be accessed via getChainInfo
+        chainInfo   *chain.Info
+        // ...
+    }
+
+    // ...
+
+    bh.chainInfoLk.RLock()
+    if bh.chainInfo != nil {
+        info := bh.chainInfo
+        bh.chainInfoLk.RUnlock()
+        return info, nil
+    }
+    bh.chainInfoLk.RUnlock()
+```
+- Some linting errors:
+```
+cmd/client/main.go:163:45: G114: Use of net/http serve function that has no support for setting timeouts (gosec)
+                        log.DefaultLogger().Fatalw("", "client", http.ListenAndServe(address, nil))
+                                                                 ^
+cmd/relay/main.go:152:9: G114: Use of net/http serve function that has no support for setting timeouts (gosec)
+        return http.Serve(listener, handler.GetHTTPHandler())
+               ^
+log/log.go:16:13: the interface has more than 10 methods: 14 (interfacebloat)
+type Logger interface {
+            ^
+cmd/drand-cli/cli_test.go:8:2: SA1019: "io/ioutil" has been deprecated since Go 1.16: As of Go 1.16, the same functionality is now provided by package io or package os, and those implementations should be preferred in new code. See the specific function documentation for details. (staticcheck)
+        "io/ioutil"
+        ^
+```
+
+- GoLand reports a few issues that might cause problems, especially around the nilness analyzer.
+- Some tests produce errors:
+
+TestLocalOrchestration
+```
+=== RUN   TestLocalOrchestration
+
+2022-10-03T22:49:32.804+0300    ERROR    127.0.0.1:44703.default.0    beacon/node.go:410        {"beacon_round": 6, "err_request": "rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing dial tcp 127.0.0.1:41087: connect: connection refused\"", "from": "127.0.0.1:41087"}
+2022-10-03T22:49:32.804+0300    ERROR    127.0.0.1:44703.default.0.SyncManager.tryNode    beacon/sync_manager.go:358    unable_to_sync    {"with_peer": "127.0.0.1:37611", "err": "rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing dial tcp 127.0.0.1:37611: connect: connection refused\""}
+```
+
+TestHTTPWaiting
+```
+=== RUN   TestHTTPWaiting
+2022-10-04T09:42:09.384+0300    INFO    http/server.go:127    New beacon handler registered    {"chainHash": "c5f1466e51ac94baa8a04fc9d5de8cd8f5e0cbb4985b70314e125d41f783b45c"}
+MOCK SERVER: emit round done
+2022-10-04T09:42:09.390+0300    WARN    http/server.go:390        {"http_server": "request in the future", "client": "[::1]:59544", "req": "%2Fc5f1466e51ac94baa8a04fc9d5de8cd8f5e0cbb4985b70314e125d41f783b45c%2Fpublic%2F1971"}
```

Comment:
This I had never seen in local tests, I'd be very interested in having a way to debug / reproduce more!

Reply:
I think I can reliably reproduce this on my machine, will try a bit more and ping you for a pairing session maybe?